contracts/tokens/Badge.sol

```diff
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.11;
+
+// Author: Francesco Sullo <francesco@superpower.io>
+// (c) Superpower Labs Inc.
+
+import "@openzeppelin/contracts-upgradeable/token/ERC721/ERC721Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/AddressUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/StringsUpgradeable.sol";
+import "soliutils/contracts/UUPSUpgradableTemplate.sol";
+
+import "../interfaces/IBadge.sol";
+
+//import "hardhat/console.sol";
+
+contract Badge is IBadge, ERC721Upgradeable, UUPSUpgradableTemplate {
+  using AddressUpgradeable for address;
+
+  string private _baseTokenURI;
+  bool private _baseTokenURIFrozen;
+  uint256 private _nextTokenId;
+
+  address public city;
+  address public minter;
+
+  mapping(uint256 => mapping(address => mapping(uint256 => uint256))) internal _tokenAttributes;
+
+  modifier onlyCity() {
+    require(city != address(0) && _msgSender() == city, "Badge: not the city");
+    _;
+  }
+
+  modifier onlyMinter() {
+    require(minter != address(0) && _msgSender() == minter, "Badge: forbidden");
+    _;
+  }
+
+  modifier tokenExists(uint256 id) {
+    require(_exists(id), "Badge: token does not exist");
+    _;
+  }
+
+  /// @custom:oz-upgrades-unsafe-allow constructor
+  constructor() initializer {}
+
+  function initialize(
+    string memory name,
+    string memory symbol,
+    string memory tokenUri
+  ) public initializer {
+    __ERC721_init(name, symbol);
+    __UUPSUpgradableTemplate_init();
+    _baseTokenURI = tokenUri;
+    _nextTokenId = 1;
+  }
+
+  function attributesOf(
+    uint256 _id,
+    address _player,
+    uint256 _index
+  ) external view override returns (uint256) {
+    return _tokenAttributes[_id][_player][_index];
+  }
+
+  function initializeAttributesFor(uint256, address) external override {
+    // we allow only the city to use the badge
+    revert("Badge: only city is authorized on this badge");
+  }
```

Comment:
Is this due to one of the dependencies requiring this function to be implemented?

Reply:
Yes